Recitation 2 (#14)
cs3157-borowski/recitations: recitation_2/makefile.md

```bash
myprogram: hello.o goodbye.o myprogram.o
	gcc -o myprogram hello.o goodbye.o myprogram.o
```

where:

- `myprogram` is the name of the target file this rule will build
- `hello.o`, `goodbye.o`, `myprogram.o` are the dependencies (prerequisites) that the target depends on
- `gcc -o myprogram hello.o goodbye.o myprogram.o` is the recipe (ie. the shell command that builds the target)
    - `goodbye.o` and `myprogram.o` are the ingredients for this recipe

Make uses the “last modified” timestamp of each file to figure out if a target needs to be rebuilt. Make will rebuild a target if it doesn’t exist, or if it was last modified earlier than one of its dependencies.

Creating our `Makefile` for `myprogram`:

```bash
myprogram: hello.o goodbye.o myprogram.o
	gcc -o myprogram hello.o goodbye.o myprogram.o

hello.o: hello.c hello.h
	gcc -c -o hello.o hello.c

goodbye.o: goodbye.c goodbye.h
	gcc -c -o goodbye.o goodbye.c

myprogram.o: myprogram.c hello.h goodbye.h
	gcc -c -o myprogram.o myprogram.c
```

> Note: The space preceding your recipe needs to be a tab, not spaces.

Reviewer: ```suggestion
Note: The space preceding your recipe needs to be a tab, not four spaces.
```

Reviewer: They look the same, but 4 spaces won't work. It's necessary to emphasize makefile is very sensitive, so be mindful when configuring vimrc.

- Replace `<filename>.o: <filename>.c <filename>.h` with `%.o: %.c %.h`

Altogether, we get this `Makefile`:

```bash
CC      = gcc
TARGET  = myprogram
C_FILES = $(wildcard *.c)
OBJS    = $(patsubst %.c,%.o,$(C_FILES))

$(TARGET): $(OBJS)
	$(CC) -o $(TARGET) $(OBJS)
    
%.o: %.c %.h
	$(CC) -c -o $@ $<
```

However, note that  `myprogram.o` doesn’t depend on `myprogram.h`, as no such file exists. 

So, we can filter out `myprogram.c` from our `C_FILES` with the **`filter-out`** function.

- Replace `C_FILES = $(wildcard *.c)` with `C_FILES = $(filter-out $(TARGET).c, $(wildcard *.c)) `

We then have to compile and link `myprogram.c` with the rest of the object files to produce `myprogram`, so we can change our rule for `myprogram` to:

```bash
$(TARGET): $(OBJS) $(TARGET).c
	$(CC) -o $(TARGET) $(OBJS) $(TARGET).c
```

This works because the command `gcc -o <output name> <ingredients>` will both compile (if needed for `.c` files) and link your ingredients.

> Note: For multi-source programs, however, it is not good practice to compile and link all your files in one step because it defeats the purpose of incremental building and individual dependencies cannot be checked. In our `Makefile`, aside from the `.c` file containing the main function, all other `.o` objects are still built separately.

Reviewer: ```suggestion
Note: For multi-source programs, however, it is not good practice to compile and link all your files in one step because it defeats the purpose of incremental building, and individual dependencies cannot be checked. In our `Makefile`, aside from the `.c` file containing the main function, all other `.o` objects are still built separately.
```

Reviewer: fanboys

```bash
$ gcc -c -g -Wall -Werror -pedantic-errors -o hello.o hello.c
```

Since our `Makefile` runs shell commands for us, we can save the flags into a macro and put it in the recipe of our rule.

Similarly, we might want to pass flags that enable options for linking.

This gives us:

```bash
CC      = gcc
TARGET  = myprogram
C_FILES = $(filter-out $(TARGET).c, $(wildcard *.c))
OBJS    = $(patsubst %.c,%.o,$(C_FILES))
CFLAGS  = -g -Wall -Werror -pedantic-errors # compiler flags
LDFLAGS = # linker flags

$(TARGET): $(OBJS) $(TARGET).c
	$(CC) -o $(TARGET) $(OBJS) $(TARGET).c $(LDFLAGS) # adding linker flags to recipe
%.o: %.c %.h
	$(CC) $(CFLAGS) -c -o $@ $< # adding compiler flags to recipe
```

**`make all` and `make clean`:**

Compiling a program is not the only thing your might want to write rules for. Makefiles commonly provide instructions on a few other things besides compiling a program: 

- To delete all the object files and executables so that the directory is ‘clean’. We can create a rule with the target `clean`, then with a recipe to `rm` the object files and executables that we made.

> - An easy-to-read `all` target that is usually the first rule in the `Makefile`. This is helpful because the `make` command will build the first target in your `Makefile`, so having the first target be `all` makes it easy to see what `make` will build.

Reviewer: ```suggestion
- An easy-to-read `all` target that is usually the first rule in the `Makefile`. This is helpful because the `make` command will build the first target in your `Makefile`, so setting the first target to `all` makes it easy to see what `make` will build.
```

Reviewer: trivial, but prefer action over the status
